{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7bd72be9-e615-48b3-9a55-af2db5f3d4ac",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T22:29:26.556329Z",
     "iopub.status.busy": "2023-06-13T22:29:26.555931Z",
     "iopub.status.idle": "2023-06-13T22:29:26.561514Z",
     "shell.execute_reply": "2023-06-13T22:29:26.560606Z",
     "shell.execute_reply.started": "2023-06-13T22:29:26.556299Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import seqdata as sd\n",
    "import xarray as xr\n",
    "import numpy as np\n",
    "import dask.array as da\n",
    "import dask_ml as dml\n",
    "from graph_part import train_test_validation_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "52f60838-a08f-4d5a-a45d-6bcb931f0a97",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T05:21:33.440314Z",
     "iopub.status.busy": "2023-06-13T05:21:33.439756Z",
     "iopub.status.idle": "2023-06-13T05:21:34.331770Z",
     "shell.execute_reply": "2023-06-13T05:21:34.330702Z",
     "shell.execute_reply.started": "2023-06-13T05:21:33.440271Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:     (_sequence: 1013080, cov_sample: 1, _length: 500)\n",
       "Coordinates:\n",
       "  * cov_sample  (cov_sample) object &#x27;stam&#x27;\n",
       "Dimensions without coordinates: _sequence, _length\n",
       "Data variables:\n",
       "    chrom       (_sequence) object dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;\n",
       "    chromEnd    (_sequence) int64 dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;\n",
       "    chromStart  (_sequence) int64 dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;\n",
       "    cov         (_sequence, cov_sample, _length) uint16 dask.array&lt;chunksize=(4096, 1, 500), meta=np.ndarray&gt;\n",
       "    seq         (_sequence, _length) |S1 dask.array&lt;chunksize=(4096, 500), meta=np.ndarray&gt;\n",
       "    strand      (_sequence) object dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;\n",
       "Attributes:\n",
       "    length_dim:    _length\n",
       "    max_jitter:    0\n",
       "    sequence_dim:  _sequence</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-79c6edf1-7687-40d0-828f-a6e86e5bb3d7' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-79c6edf1-7687-40d0-828f-a6e86e5bb3d7' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>_sequence</span>: 1013080</li><li><span class='xr-has-index'>cov_sample</span>: 1</li><li><span>_length</span>: 500</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-e3b3f7ee-fbf0-4ce5-9e56-b19b7f42c21d' class='xr-section-summary-in' type='checkbox'  checked><label for='section-e3b3f7ee-fbf0-4ce5-9e56-b19b7f42c21d' class='xr-section-summary' >Coordinates: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cov_sample</span></div><div class='xr-var-dims'>(cov_sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;stam&#x27;</div><input id='attrs-67317d14-1083-459f-a92a-7c55ab8ab304' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-67317d14-1083-459f-a92a-7c55ab8ab304' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2376af56-00b5-476d-8572-96e462b8bb15' class='xr-var-data-in' type='checkbox'><label for='data-2376af56-00b5-476d-8572-96e462b8bb15' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><pre>array([&#x27;stam&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-9f839bd2-4f2a-4e71-9d81-fa539c83eea6' class='xr-section-summary-in' type='checkbox'  checked><label for='section-9f839bd2-4f2a-4e71-9d81-fa539c83eea6' class='xr-section-summary' >Data variables: <span>(6)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>chrom</span></div><div class='xr-var-dims'>(_sequence)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;</div><input id='attrs-66b9c785-2735-438a-8936-8f6b2ab964f9' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-66b9c785-2735-438a-8936-8f6b2ab964f9' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-111dbd9b-1dae-4fb9-82ca-c017f0fe3048' class='xr-var-data-in' type='checkbox'><label for='data-111dbd9b-1dae-4fb9-82ca-c017f0fe3048' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 7.73 MiB </td>\n",
       "                        <td> 494.67 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080,) </td>\n",
       "                        <td> (63318,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 16 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> object numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"7\" y1=\"0\" x2=\"7\" y2=\"25\" />\n",
       "  <line x1=\"15\" y1=\"0\" x2=\"15\" y2=\"25\" />\n",
       "  <line x1=\"22\" y1=\"0\" x2=\"22\" y2=\"25\" />\n",
       "  <line x1=\"30\" y1=\"0\" x2=\"30\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"45\" y1=\"0\" x2=\"45\" y2=\"25\" />\n",
       "  <line x1=\"52\" y1=\"0\" x2=\"52\" y2=\"25\" />\n",
       "  <line x1=\"60\" y1=\"0\" x2=\"60\" y2=\"25\" />\n",
       "  <line x1=\"67\" y1=\"0\" x2=\"67\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"82\" y1=\"0\" x2=\"82\" y2=\"25\" />\n",
       "  <line x1=\"90\" y1=\"0\" x2=\"90\" y2=\"25\" />\n",
       "  <line x1=\"97\" y1=\"0\" x2=\"97\" y2=\"25\" />\n",
       "  <line x1=\"105\" y1=\"0\" x2=\"105\" y2=\"25\" />\n",
       "  <line x1=\"112\" y1=\"0\" x2=\"112\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1013080</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>chromEnd</span></div><div class='xr-var-dims'>(_sequence)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;</div><input id='attrs-75f550ed-a171-4ef8-b609-942e1169b600' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-75f550ed-a171-4ef8-b609-942e1169b600' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a4a8eef3-fcab-4c0a-bc5d-8cd507c6a515' class='xr-var-data-in' type='checkbox'><label for='data-a4a8eef3-fcab-4c0a-bc5d-8cd507c6a515' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 7.73 MiB </td>\n",
       "                        <td> 494.67 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080,) </td>\n",
       "                        <td> (63318,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 16 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> int64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"7\" y1=\"0\" x2=\"7\" y2=\"25\" />\n",
       "  <line x1=\"15\" y1=\"0\" x2=\"15\" y2=\"25\" />\n",
       "  <line x1=\"22\" y1=\"0\" x2=\"22\" y2=\"25\" />\n",
       "  <line x1=\"30\" y1=\"0\" x2=\"30\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"45\" y1=\"0\" x2=\"45\" y2=\"25\" />\n",
       "  <line x1=\"52\" y1=\"0\" x2=\"52\" y2=\"25\" />\n",
       "  <line x1=\"60\" y1=\"0\" x2=\"60\" y2=\"25\" />\n",
       "  <line x1=\"67\" y1=\"0\" x2=\"67\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"82\" y1=\"0\" x2=\"82\" y2=\"25\" />\n",
       "  <line x1=\"90\" y1=\"0\" x2=\"90\" y2=\"25\" />\n",
       "  <line x1=\"97\" y1=\"0\" x2=\"97\" y2=\"25\" />\n",
       "  <line x1=\"105\" y1=\"0\" x2=\"105\" y2=\"25\" />\n",
       "  <line x1=\"112\" y1=\"0\" x2=\"112\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1013080</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>chromStart</span></div><div class='xr-var-dims'>(_sequence)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;</div><input id='attrs-2401e4a7-ca81-4911-b691-f5609014d953' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-2401e4a7-ca81-4911-b691-f5609014d953' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-1c3d1600-3e13-42a7-99f8-bdfc7bcbad60' class='xr-var-data-in' type='checkbox'><label for='data-1c3d1600-3e13-42a7-99f8-bdfc7bcbad60' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 7.73 MiB </td>\n",
       "                        <td> 494.67 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080,) </td>\n",
       "                        <td> (63318,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 16 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> int64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"7\" y1=\"0\" x2=\"7\" y2=\"25\" />\n",
       "  <line x1=\"15\" y1=\"0\" x2=\"15\" y2=\"25\" />\n",
       "  <line x1=\"22\" y1=\"0\" x2=\"22\" y2=\"25\" />\n",
       "  <line x1=\"30\" y1=\"0\" x2=\"30\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"45\" y1=\"0\" x2=\"45\" y2=\"25\" />\n",
       "  <line x1=\"52\" y1=\"0\" x2=\"52\" y2=\"25\" />\n",
       "  <line x1=\"60\" y1=\"0\" x2=\"60\" y2=\"25\" />\n",
       "  <line x1=\"67\" y1=\"0\" x2=\"67\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"82\" y1=\"0\" x2=\"82\" y2=\"25\" />\n",
       "  <line x1=\"90\" y1=\"0\" x2=\"90\" y2=\"25\" />\n",
       "  <line x1=\"97\" y1=\"0\" x2=\"97\" y2=\"25\" />\n",
       "  <line x1=\"105\" y1=\"0\" x2=\"105\" y2=\"25\" />\n",
       "  <line x1=\"112\" y1=\"0\" x2=\"112\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1013080</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>cov</span></div><div class='xr-var-dims'>(_sequence, cov_sample, _length)</div><div class='xr-var-dtype'>uint16</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(4096, 1, 500), meta=np.ndarray&gt;</div><input id='attrs-5decd224-b23e-43f8-b773-bcfec9614d81' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-5decd224-b23e-43f8-b773-bcfec9614d81' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-12d6543f-e946-4fe7-82b5-581e7855f30b' class='xr-var-data-in' type='checkbox'><label for='data-12d6543f-e946-4fe7-82b5-581e7855f30b' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 0.94 GiB </td>\n",
       "                        <td> 3.91 MiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080, 1, 500) </td>\n",
       "                        <td> (4096, 1, 500) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 248 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint16 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"156\" height=\"146\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"10\" y1=\"0\" x2=\"80\" y2=\"70\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"10\" y1=\"25\" x2=\"80\" y2=\"96\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"10\" y1=\"0\" x2=\"10\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"13\" y1=\"3\" x2=\"13\" y2=\"29\" />\n",
       "  <line x1=\"17\" y1=\"7\" x2=\"17\" y2=\"32\" />\n",
       "  <line x1=\"21\" y1=\"11\" x2=\"21\" y2=\"36\" />\n",
       "  <line x1=\"24\" y1=\"14\" x2=\"24\" y2=\"40\" />\n",
       "  <line x1=\"28\" y1=\"18\" x2=\"28\" y2=\"43\" />\n",
       "  <line x1=\"32\" y1=\"22\" x2=\"32\" y2=\"47\" />\n",
       "  <line x1=\"35\" y1=\"25\" x2=\"35\" y2=\"51\" />\n",
       "  <line x1=\"39\" y1=\"29\" x2=\"39\" y2=\"55\" />\n",
       "  <line x1=\"43\" y1=\"33\" x2=\"43\" y2=\"58\" />\n",
       "  <line x1=\"47\" y1=\"37\" x2=\"47\" y2=\"62\" />\n",
       "  <line x1=\"50\" y1=\"40\" x2=\"50\" y2=\"66\" />\n",
       "  <line x1=\"54\" y1=\"44\" x2=\"54\" y2=\"69\" />\n",
       "  <line x1=\"58\" y1=\"48\" x2=\"58\" y2=\"73\" />\n",
       "  <line x1=\"61\" y1=\"51\" x2=\"61\" y2=\"77\" />\n",
       "  <line x1=\"65\" y1=\"55\" x2=\"65\" y2=\"81\" />\n",
       "  <line x1=\"69\" y1=\"59\" x2=\"69\" y2=\"84\" />\n",
       "  <line x1=\"73\" y1=\"63\" x2=\"73\" y2=\"88\" />\n",
       "  <line x1=\"76\" y1=\"66\" x2=\"76\" y2=\"92\" />\n",
       "  <line x1=\"80\" y1=\"70\" x2=\"80\" y2=\"96\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"10.0,0.0 80.58823529411765,70.58823529411765 80.58823529411765,96.00085180870013 10.0,25.412616514582485\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"10\" y1=\"0\" x2=\"35\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"13\" y1=\"3\" x2=\"39\" y2=\"3\" />\n",
       "  <line x1=\"17\" y1=\"7\" x2=\"42\" y2=\"7\" />\n",
       "  <line x1=\"21\" y1=\"11\" x2=\"46\" y2=\"11\" />\n",
       "  <line x1=\"24\" y1=\"14\" x2=\"50\" y2=\"14\" />\n",
       "  <line x1=\"28\" y1=\"18\" x2=\"53\" y2=\"18\" />\n",
       "  <line x1=\"32\" y1=\"22\" x2=\"57\" y2=\"22\" />\n",
       "  <line x1=\"35\" y1=\"25\" x2=\"61\" y2=\"25\" />\n",
       "  <line x1=\"39\" y1=\"29\" x2=\"65\" y2=\"29\" />\n",
       "  <line x1=\"43\" y1=\"33\" x2=\"68\" y2=\"33\" />\n",
       "  <line x1=\"47\" y1=\"37\" x2=\"72\" y2=\"37\" />\n",
       "  <line x1=\"50\" y1=\"40\" x2=\"76\" y2=\"40\" />\n",
       "  <line x1=\"54\" y1=\"44\" x2=\"79\" y2=\"44\" />\n",
       "  <line x1=\"58\" y1=\"48\" x2=\"83\" y2=\"48\" />\n",
       "  <line x1=\"61\" y1=\"51\" x2=\"87\" y2=\"51\" />\n",
       "  <line x1=\"65\" y1=\"55\" x2=\"91\" y2=\"55\" />\n",
       "  <line x1=\"69\" y1=\"59\" x2=\"94\" y2=\"59\" />\n",
       "  <line x1=\"73\" y1=\"63\" x2=\"98\" y2=\"63\" />\n",
       "  <line x1=\"76\" y1=\"66\" x2=\"102\" y2=\"66\" />\n",
       "  <line x1=\"80\" y1=\"70\" x2=\"106\" y2=\"70\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"10\" y1=\"0\" x2=\"80\" y2=\"70\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"35\" y1=\"0\" x2=\"106\" y2=\"70\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"10.0,0.0 35.41261651458248,0.0 106.00085180870013,70.58823529411765 80.58823529411765,70.58823529411765\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"80\" y1=\"70\" x2=\"106\" y2=\"70\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"80\" y1=\"96\" x2=\"106\" y2=\"96\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"80\" y1=\"70\" x2=\"80\" y2=\"96\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"106\" y1=\"70\" x2=\"106\" y2=\"96\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"80.58823529411765,70.58823529411765 106.00085180870013,70.58823529411765 106.00085180870013,96.00085180870013 80.58823529411765,96.00085180870013\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"93.294544\" y=\"116.000852\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >500</text>\n",
       "  <text x=\"126.000852\" y=\"83.294544\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,126.000852,83.294544)\">1</text>\n",
       "  <text x=\"35.294118\" y=\"80.706734\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(45,35.294118,80.706734)\">1013080</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>seq</span></div><div class='xr-var-dims'>(_sequence, _length)</div><div class='xr-var-dtype'>|S1</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(4096, 500), meta=np.ndarray&gt;</div><input id='attrs-e4599e22-b2e9-4f0a-beaf-421f7090bc70' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e4599e22-b2e9-4f0a-beaf-421f7090bc70' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-2c6244e2-89f5-46f4-9b1f-093d4d836724' class='xr-var-data-in' type='checkbox'><label for='data-2c6244e2-89f5-46f4-9b1f-093d4d836724' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 483.07 MiB </td>\n",
       "                        <td> 1.95 MiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080, 500) </td>\n",
       "                        <td> (4096, 500) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 248 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> |S1 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"75\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"25\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"6\" x2=\"25\" y2=\"6\" />\n",
       "  <line x1=\"0\" y1=\"12\" x2=\"25\" y2=\"12\" />\n",
       "  <line x1=\"0\" y1=\"18\" x2=\"25\" y2=\"18\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"25\" y2=\"25\" />\n",
       "  <line x1=\"0\" y1=\"31\" x2=\"25\" y2=\"31\" />\n",
       "  <line x1=\"0\" y1=\"37\" x2=\"25\" y2=\"37\" />\n",
       "  <line x1=\"0\" y1=\"44\" x2=\"25\" y2=\"44\" />\n",
       "  <line x1=\"0\" y1=\"50\" x2=\"25\" y2=\"50\" />\n",
       "  <line x1=\"0\" y1=\"56\" x2=\"25\" y2=\"56\" />\n",
       "  <line x1=\"0\" y1=\"63\" x2=\"25\" y2=\"63\" />\n",
       "  <line x1=\"0\" y1=\"69\" x2=\"25\" y2=\"69\" />\n",
       "  <line x1=\"0\" y1=\"75\" x2=\"25\" y2=\"75\" />\n",
       "  <line x1=\"0\" y1=\"81\" x2=\"25\" y2=\"81\" />\n",
       "  <line x1=\"0\" y1=\"88\" x2=\"25\" y2=\"88\" />\n",
       "  <line x1=\"0\" y1=\"94\" x2=\"25\" y2=\"94\" />\n",
       "  <line x1=\"0\" y1=\"100\" x2=\"25\" y2=\"100\" />\n",
       "  <line x1=\"0\" y1=\"107\" x2=\"25\" y2=\"107\" />\n",
       "  <line x1=\"0\" y1=\"113\" x2=\"25\" y2=\"113\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"25\" y1=\"0\" x2=\"25\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 25.412616514582485,0.0 25.412616514582485,120.0 0.0,120.0\" style=\"fill:#8B4903A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"12.706308\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >500</text>\n",
       "  <text x=\"45.412617\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(-90,45.412617,60.000000)\">1013080</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>strand</span></div><div class='xr-var-dims'>(_sequence)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(63318,), meta=np.ndarray&gt;</div><input id='attrs-3f6cd600-00eb-4fc4-ad8d-90e3e472c8b3' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-3f6cd600-00eb-4fc4-ad8d-90e3e472c8b3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9b3622af-5447-4e91-abb8-5822356cf87f' class='xr-var-data-in' type='checkbox'><label for='data-9b3622af-5447-4e91-abb8-5822356cf87f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 7.73 MiB </td>\n",
       "                        <td> 494.67 kiB </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1013080,) </td>\n",
       "                        <td> (63318,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 16 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> object numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"75\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"25\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"7\" y1=\"0\" x2=\"7\" y2=\"25\" />\n",
       "  <line x1=\"15\" y1=\"0\" x2=\"15\" y2=\"25\" />\n",
       "  <line x1=\"22\" y1=\"0\" x2=\"22\" y2=\"25\" />\n",
       "  <line x1=\"30\" y1=\"0\" x2=\"30\" y2=\"25\" />\n",
       "  <line x1=\"37\" y1=\"0\" x2=\"37\" y2=\"25\" />\n",
       "  <line x1=\"45\" y1=\"0\" x2=\"45\" y2=\"25\" />\n",
       "  <line x1=\"52\" y1=\"0\" x2=\"52\" y2=\"25\" />\n",
       "  <line x1=\"60\" y1=\"0\" x2=\"60\" y2=\"25\" />\n",
       "  <line x1=\"67\" y1=\"0\" x2=\"67\" y2=\"25\" />\n",
       "  <line x1=\"75\" y1=\"0\" x2=\"75\" y2=\"25\" />\n",
       "  <line x1=\"82\" y1=\"0\" x2=\"82\" y2=\"25\" />\n",
       "  <line x1=\"90\" y1=\"0\" x2=\"90\" y2=\"25\" />\n",
       "  <line x1=\"97\" y1=\"0\" x2=\"97\" y2=\"25\" />\n",
       "  <line x1=\"105\" y1=\"0\" x2=\"105\" y2=\"25\" />\n",
       "  <line x1=\"112\" y1=\"0\" x2=\"112\" y2=\"25\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"25\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,25.412616514582485 0.0,25.412616514582485\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"45.412617\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1013080</text>\n",
       "  <text x=\"140.000000\" y=\"12.706308\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,12.706308)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-8d00e359-44c6-47d0-a627-2d065747fb7d' class='xr-section-summary-in' type='checkbox'  ><label for='section-8d00e359-44c6-47d0-a627-2d065747fb7d' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>cov_sample</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-945c8f42-8f14-46b4-b802-aadb0d668ecc' class='xr-index-data-in' type='checkbox'/><label for='index-945c8f42-8f14-46b4-b802-aadb0d668ecc' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;stam&#x27;], dtype=&#x27;object&#x27;, name=&#x27;cov_sample&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-1a07f85b-b381-4388-bcd7-367b08e5d427' class='xr-section-summary-in' type='checkbox'  checked><label for='section-1a07f85b-b381-4388-bcd7-367b08e5d427' class='xr-section-summary' >Attributes: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>length_dim :</span></dt><dd>_length</dd><dt><span>max_jitter :</span></dt><dd>0</dd><dt><span>sequence_dim :</span></dt><dd>_sequence</dd></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:     (_sequence: 1013080, cov_sample: 1, _length: 500)\n",
       "Coordinates:\n",
       "  * cov_sample  (cov_sample) object 'stam'\n",
       "Dimensions without coordinates: _sequence, _length\n",
       "Data variables:\n",
       "    chrom       (_sequence) object dask.array<chunksize=(63318,), meta=np.ndarray>\n",
       "    chromEnd    (_sequence) int64 dask.array<chunksize=(63318,), meta=np.ndarray>\n",
       "    chromStart  (_sequence) int64 dask.array<chunksize=(63318,), meta=np.ndarray>\n",
       "    cov         (_sequence, cov_sample, _length) uint16 dask.array<chunksize=(4096, 1, 500), meta=np.ndarray>\n",
       "    seq         (_sequence, _length) |S1 dask.array<chunksize=(4096, 500), meta=np.ndarray>\n",
       "    strand      (_sequence) object dask.array<chunksize=(63318,), meta=np.ndarray>\n",
       "Attributes:\n",
       "    length_dim:    _length\n",
       "    max_jitter:    0\n",
       "    sequence_dim:  _sequence"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sdata = sd.open_zarr('/cellar/users/dlaub/projects/ML4GLand/use_cases/kopp21/kopp21.zarr')\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f8cd13f5-6536-4c84-81ad-cd945417169a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T19:46:04.011730Z",
     "iopub.status.busy": "2023-06-13T19:46:04.011388Z",
     "iopub.status.idle": "2023-06-13T19:46:04.016849Z",
     "shell.execute_reply": "2023-06-13T19:46:04.015788Z",
     "shell.execute_reply.started": "2023-06-13T19:46:04.011702Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def train_test_split_chrom(sdata: xr.Dataset, test_chroms: list[str], train_key=None):\n",
    "    test_mask = sdata.chrom.isin(test_chroms).compute()\n",
    "    return sdata.assign()\n",
    "    return sdata.sel(_sequence=~test_mask), sdata.sel(_sequence=test_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2884652b-4944-420b-b492-a13a2c9c7e74",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T19:46:37.168007Z",
     "iopub.status.busy": "2023-06-13T19:46:37.167551Z",
     "iopub.status.idle": "2023-06-13T19:46:38.012093Z",
     "shell.execute_reply": "2023-06-13T19:46:38.011183Z",
     "shell.execute_reply.started": "2023-06-13T19:46:37.167975Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[875547, 137533]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splits = train_test_split_chrom(sdata, ['chr2', 'chr3'])\n",
    "[s.sizes['_sequence'] for s in splits]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "ca849a6b-ee43-4fea-9a4b-5126fb56246c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T19:56:15.546841Z",
     "iopub.status.busy": "2023-06-13T19:56:15.546553Z",
     "iopub.status.idle": "2023-06-13T19:56:15.552925Z",
     "shell.execute_reply": "2023-06-13T19:56:15.551865Z",
     "shell.execute_reply.started": "2023-06-13T19:56:15.546815Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def train_test_split_random(sdata: xr.Dataset, dim: str, groups=None, test_size=0.1, random_state=None):\n",
    "    splitter = dml.model_selection.ShuffleSplit(\n",
    "        n_splits=1,\n",
    "        test_size=test_size,\n",
    "        random_state=random_state\n",
    "    )\n",
    "    train_idx, test_idx = next(splitter.split(da.arange(sdata.sizes[dim]), groups=groups))\n",
    "    return sdata.isel({dim: train_idx}), sdata.isel({dim: test_idx})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "3d9e779d-899d-4f65-972f-b767b1f5fd47",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T19:59:38.668522Z",
     "iopub.status.busy": "2023-06-13T19:59:38.668059Z",
     "iopub.status.idle": "2023-06-13T19:59:38.705105Z",
     "shell.execute_reply": "2023-06-13T19:59:38.704171Z",
     "shell.execute_reply.started": "2023-06-13T19:59:38.668492Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[911772, 101308]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splits = train_test_split_random(sdata, '_sequence', sdata.chrom)\n",
    "[s.sizes['_sequence'] for s in splits]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "0f1a627c-1a1a-4cb9-9d3a-3b55d531fde3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-06-13T22:31:27.156162Z",
     "iopub.status.busy": "2023-06-13T22:31:27.155768Z",
     "iopub.status.idle": "2023-06-13T22:31:27.162971Z",
     "shell.execute_reply": "2023-06-13T22:31:27.162084Z",
     "shell.execute_reply.started": "2023-06-13T22:31:27.156132Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def train_test_split_homology(sdata: xr.Dataset, seq_var: str, test_size=0.1, nucleotide=True):\n",
    "    seq_length = sdata.sizes[sdata.attrs['length_dim']]\n",
    "    outs = train_test_validation_split(\n",
    "        sdata[seq_var].to_numpy().view(f'S{seq_length}').squeeze().astype('U').astype(object),\n",
    "        test_size=test_size,\n",
    "        initialization_mode='fast-nn',\n",
    "        nucleotide=nucleotide,\n",
    "        prefilter=True,\n",
    "        denominator='shortest'\n",
    "    )\n",
    "    train_idx, test_idx = map(np.array, outs)\n",
    "    return sdata.isel({sdata.attrs['sequence_dim']: train_idx}), sdata.isel({sdata.attrs['sequence_dim']: test_idx})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c578b1e-b882-462a-ae25-b2ac56735327",
   "metadata": {},
   "source": [
    "Splitting by homology with 10,000 sequences takes ~6 minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "10cf8f62-cbdd-43ef-a0da-242e2748f021",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2023-06-13T21:24:27.012676Z",
     "iopub.status.busy": "2023-06-13T21:24:27.012311Z",
     "iopub.status.idle": "2023-06-13T21:30:49.198572Z",
     "shell.execute_reply": "2023-06-13T21:30:49.197274Z",
     "shell.execute_reply.started": "2023-06-13T21:24:27.012636Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "temp/seq_db exists and will be overwritten\n",
      "createdb --dbtype 2 graphpart_api.fasta.tmp temp/seq_db \n",
      "\n",
      "MMseqs Version:       \t14.7e284\n",
      "Database type         \t2\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to seq_db_h: 0h 0m 0s 99ms\n",
      "Time for merging to seq_db: 0h 0m 0s 117ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 0s 612ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "prefilter -s 7.5 temp/seq_db temp/seq_db temp/pref \n",
      "\n",
      "MMseqs Version:           \t14.7e284\n",
      "Substitution matrix       \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix  \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity               \t7.5\n",
      "k-mer length              \t0\n",
      "k-score                   \tseq:2147483647,prof:2147483647\n",
      "Alphabet size             \taa:21,nucl:5\n",
      "Max sequence length       \t65535\n",
      "Max results per query     \t300\n",
      "Split database            \t0\n",
      "Split mode                \t2\n",
      "Split memory limit        \t0\n",
      "Coverage threshold        \t0\n",
      "Coverage mode             \t0\n",
      "Compositional bias        \t1\n",
      "Compositional bias        \t1\n",
      "Diagonal scoring          \ttrue\n",
      "Exact k-mer matching      \t0\n",
      "Mask residues             \t1\n",
      "Mask residues probability \t0.9\n",
      "Mask lower case residues  \t0\n",
      "Minimum diagonal score    \t15\n",
      "Selected taxa             \t\n",
      "Include identical seq. id.\tfalse\n",
      "Spaced k-mers             \t1\n",
      "Preload mode              \t0\n",
      "Pseudo count a            \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b            \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern      \t\n",
      "Local temporary path      \t\n",
      "Threads                   \t96\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 10000 type: Nucleotide\n",
      "Estimated memory consumption: 77M\n",
      "Target database size: 10000 type: Nucleotide\n",
      "Index table k-mer threshold: 0 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 10.00K 0s 859ms\n",
      "Index table: Masked residues: 155503\n",
      "Index table: fill\n",
      "[=================================================================] 10.00K 0s 338ms\n",
      "Index statistics\n",
      "Entries:          4233450\n",
      "DB size:          24 MB\n",
      "Avg k-mer size:   1033.557129\n",
      "Top 10 k-mers\n",
      "    CCGCGG\t3950\n",
      "    CCAGGG\t3291\n",
      "    GGGGGG\t3227\n",
      "    TGATAG\t3095\n",
      "    GCGGTG\t3016\n",
      "    GCGGGG\t2892\n",
      "    CTCGCT\t2809\n",
      "    CCCCAG\t2803\n",
      "    GGATAG\t2787\n",
      "    GGGTAG\t2731\n",
      "Time for index table init: 0h 0m 1s 587ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 0\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 10000\n",
      "Target db start 1 to 10000\n",
      "[=================================================================] 10.00K 43s 508ms\n",
      "\n",
      "0.982000 k-mers per position\n",
      "657477 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "300 sequences passed prefiltering per query sequence\n",
      "300 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref: 0h 0m 0s 181ms\n",
      "Time for processing: 0h 0m 46s 714ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "align temp/seq_db temp/seq_db temp/pref temp/align_db --alignment-mode 3 -e inf --seq-id-mode 1 --min-seq-id 0.3 \n",
      "\n",
      "MMseqs Version:           \t14.7e284\n",
      "Substitution matrix       \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace             \tfalse\n",
      "Alignment mode            \t3\n",
      "Alignment mode            \t0\n",
      "Allow wrapped scoring     \tfalse\n",
      "E-value threshold         \tinf\n",
      "Seq. id. threshold        \t0.3\n",
      "Min alignment length      \t0\n",
      "Seq. id. mode             \t1\n",
      "Alternative alignments    \t0\n",
      "Coverage threshold        \t0\n",
      "Coverage mode             \t0\n",
      "Max sequence length       \t65535\n",
      "Compositional bias        \t1\n",
      "Compositional bias        \t1\n",
      "Max reject                \t2147483647\n",
      "Max accept                \t2147483647\n",
      "Include identical seq. id.\tfalse\n",
      "Preload mode              \t0\n",
      "Pseudo count a            \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b            \tsubstitution:4.100,context:5.800\n",
      "Score bias                \t0\n",
      "Realign hits              \tfalse\n",
      "Realign score bias        \t-0.2\n",
      "Realign max seqs          \t2147483647\n",
      "Correlation score weight  \t0\n",
      "Gap open cost             \taa:11,nucl:5\n",
      "Gap extension cost        \taa:1,nucl:2\n",
      "Zdrop                     \t40\n",
      "Threads                   \t96\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 10000 type: Nucleotide\n",
      "Target database size: 10000 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 10.00K 1m 15s 796ms\n",
      "Time for merging to align_db: 0h 0m 0s 159ms\n",
      "3000000 alignments calculated\n",
      "694815 sequence pairs passed the thresholds (0.231605 of overall calculated)\n",
      "69.481499 hits per query sequence\n",
      "Time for processing: 0h 1m 17s 47ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "convertalis temp/seq_db temp/seq_db temp/align_db temp/alignments.tab \n",
      "\n",
      "MMseqs Version:        \t14.7e284\n",
      "Substitution matrix    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Alignment format       \t0\n",
      "Format alignment output\tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits\n",
      "Translation table      \t1\n",
      "Gap open cost          \taa:11,nucl:5\n",
      "Gap extension cost     \taa:1,nucl:2\n",
      "Database output        \tfalse\n",
      "Preload mode           \t0\n",
      "Search type            \t0\n",
      "Threads                \t96\n",
      "Compressed             \t0\n",
      "Verbosity              \t3\n",
      "\n",
      "[=================================================================] 10.00K 0s 979ms\n",
      "Time for merging to alignments.tab: 0h 0m 0s 421ms\n",
      "Time for processing: 0h 0m 2s 150ms\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "91d85a0c0f824fe58cf42f1c22ee7b6a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full graph nr. of edges: 488825\n",
      "1000 [1000]\n",
      "Initialization mode fast-nn\n",
      "1\n",
      "Currently have this many samples: 10000\n",
      "!  seq_0 seq_7030 {'metric': 0.544}  !\n",
      "Need to remove! Currently have this many samples: 10000\n",
      "After removal we have this many samples: 9867\n"
     ]
    }
   ],
   "source": [
    "train, test = train_test_split_homology(sdata.isel(_sequence=slice(0, int(1e4))), 'seq')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d589077f-2a88-4f89-b760-9a1c4cda460e",
   "metadata": {},
   "source": [
    "Note: we don't demonstrate this for more than 100,000 sequences since this takes ~3.5 hours to run. The homology graph partitioning algorithm needs approximately $O(n^2)$ time so increasing the amount of sequences beyond this would be intractable for the purposes of a tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "dbd8225e-e0ab-430e-b17b-3fa7a5870605",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2023-06-13T22:31:54.374933Z",
     "iopub.status.busy": "2023-06-13T22:31:54.374481Z",
     "iopub.status.idle": "2023-06-14T02:01:29.649678Z",
     "shell.execute_reply": "2023-06-14T02:01:29.648250Z",
     "shell.execute_reply.started": "2023-06-13T22:31:54.374903Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "createdb --dbtype 2 graphpart_api.fasta.tmp temp/seq_db \n",
      "\n",
      "MMseqs Version:       \t14.7e284\n",
      "Database type         \t2\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[=========\n",
      "Time for merging to seq_db_h: 0h 0m 0s 182ms\n",
      "Time for merging to seq_db: 0h 0m 0s 345ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 2s 966ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "prefilter -s 7.5 temp/seq_db temp/seq_db temp/pref \n",
      "\n",
      "MMseqs Version:           \t14.7e284\n",
      "Substitution matrix       \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix  \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity               \t7.5\n",
      "k-mer length              \t0\n",
      "k-score                   \tseq:2147483647,prof:2147483647\n",
      "Alphabet size             \taa:21,nucl:5\n",
      "Max sequence length       \t65535\n",
      "Max results per query     \t300\n",
      "Split database            \t0\n",
      "Split mode                \t2\n",
      "Split memory limit        \t0\n",
      "Coverage threshold        \t0\n",
      "Coverage mode             \t0\n",
      "Compositional bias        \t1\n",
      "Compositional bias        \t1\n",
      "Diagonal scoring          \ttrue\n",
      "Exact k-mer matching      \t0\n",
      "Mask residues             \t1\n",
      "Mask residues probability \t0.9\n",
      "Mask lower case residues  \t0\n",
      "Minimum diagonal score    \t15\n",
      "Selected taxa             \t\n",
      "Include identical seq. id.\tfalse\n",
      "Spaced k-mers             \t1\n",
      "Preload mode              \t0\n",
      "Pseudo count a            \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b            \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern      \t\n",
      "Local temporary path      \t\n",
      "Threads                   \t96\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Query database size: 100000 type: Nucleotide\n",
      "Estimated memory consumption: 769M\n",
      "Target database size: 100000 type: Nucleotide\n",
      "Index table k-mer threshold: 0 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 100.00K 6s 112ms\n",
      "Index table: Masked residues: 1051717\n",
      "Index table: fill\n",
      "[=================================================================] 100.00K 2s 530ms\n",
      "Index statistics\n",
      "Entries:          43087889\n",
      "DB size:          246 MB\n",
      "Avg k-mer size:   10519.504150\n",
      "Top 10 k-mers\n",
      "    AAAAAA\t39562\n",
      "    TTTTTT\t39004\n",
      "    TGATAG\t29432\n",
      "    AAAATT\t28059\n",
      "    ATTTTT\t28004\n",
      "    TTTATT\t27286\n",
      "    CCGCGG\t27188\n",
      "    CTTTTT\t26157\n",
      "    AAAAAG\t26148\n",
      "    GGATAG\t25507\n",
      "Time for index table init: 0h 0m 10s 159ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 0\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 100000\n",
      "Target db start 1 to 100000\n",
      "[=================================================================] 100.00K 1h 12m 46s 283ms\n",
      "\n",
      "0.982000 k-mers per position\n",
      "6548001 DB matches per sequence\n",
      "100000 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "300 sequences passed prefiltering per query sequence\n",
      "300 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref: 0h 0m 1s 392ms\n",
      "Time for processing: 1h 13m 0s 974ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "align temp/seq_db temp/seq_db temp/pref temp/align_db --alignment-mode 3 -e inf --seq-id-mode 1 --min-seq-id 0.3 \n",
      "\n",
      "MMseqs Version:           \t14.7e284\n",
      "Substitution matrix       \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace             \tfalse\n",
      "Alignment mode            \t3\n",
      "Alignment mode            \t0\n",
      "Allow wrapped scoring     \tfalse\n",
      "E-value threshold         \tinf\n",
      "Seq. id. threshold        \t0.3\n",
      "Min alignment length      \t0\n",
      "Seq. id. mode             \t1\n",
      "Alternative alignments    \t0\n",
      "Coverage threshold        \t0\n",
      "Coverage mode             \t0\n",
      "Max sequence length       \t65535\n",
      "Compositional bias        \t1\n",
      "Compositional bias        \t1\n",
      "Max reject                \t2147483647\n",
      "Max accept                \t2147483647\n",
      "Include identical seq. id.\tfalse\n",
      "Preload mode              \t0\n",
      "Pseudo count a            \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b            \tsubstitution:4.100,context:5.800\n",
      "Score bias                \t0\n",
      "Realign hits              \tfalse\n",
      "Realign score bias        \t-0.2\n",
      "Realign max seqs          \t2147483647\n",
      "Correlation score weight  \t0\n",
      "Gap open cost             \taa:11,nucl:5\n",
      "Gap extension cost        \taa:1,nucl:2\n",
      "Zdrop                     \t40\n",
      "Threads                   \t96\n",
      "Compressed                \t0\n",
      "Verbosity                 \t3\n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 100000 type: Nucleotide\n",
      "Target database size: 100000 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 100.00K 13m 7s 632ms\n",
      "Time for merging to align_db: 0h 0m 1s 276ms\n",
      "30000000 alignments calculated\n",
      "6160388 sequence pairs passed the thresholds (0.205346 of overall calculated)\n",
      "61.603882 hits per query sequence\n",
      "Time for processing: 0h 13m 11s 169ms\n",
      "Unable to join threads to shut down before fork(). This can break multithreading in child processes.\n",
      "\n",
      "convertalis temp/seq_db temp/seq_db temp/align_db temp/alignments.tab \n",
      "\n",
      "MMseqs Version:        \t14.7e284\n",
      "Substitution matrix    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Alignment format       \t0\n",
      "Format alignment output\tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits\n",
      "Translation table      \t1\n",
      "Gap open cost          \taa:11,nucl:5\n",
      "Gap extension cost     \taa:1,nucl:2\n",
      "Database output        \tfalse\n",
      "Preload mode           \t0\n",
      "Search type            \t0\n",
      "Threads                \t96\n",
      "Compressed             \t0\n",
      "Verbosity              \t3\n",
      "\n",
      "[=================================================================] 100.00K 8s 65ms\n",
      "Time for merging to alignments.tab: 0h 0m 2s 632ms\n",
      "Time for processing: 0h 0m 14s 760ms\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4a94beec86c64fae85bc40acddfbeb91",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Full graph nr. of edges: 4977945\n",
      "10000 [10000]\n",
      "Initialization mode fast-nn\n",
      "edges: 10000\n",
      "{'cluster': 3272, 'C-size': 8085, 'label-counts': array([8085])} {'metric': 0.4}\n",
      "edges: 20000\n",
      "{'cluster': 14429, 'C-size': 75, 'label-counts': array([75])} {'metric': 0.4}\n",
      "edges: 30000\n",
      "{'cluster': 25934, 'C-size': 11, 'label-counts': array([11])} {'metric': 0.4}\n",
      "edges: 40000\n",
      "{'cluster': 36944, 'C-size': 94, 'label-counts': array([94])} {'metric': 0.4}\n",
      "edges: 50000\n",
      "{'cluster': 47984, 'C-size': 10, 'label-counts': array([10])} {'metric': 0.4}\n",
      "edges: 60000\n",
      "{'cluster': 59133, 'C-size': 8, 'label-counts': array([8])} {'metric': 0.4}\n",
      "edges: 70000\n",
      "{'cluster': 70268, 'C-size': 57, 'label-counts': array([57])} {'metric': 0.4}\n",
      "edges: 80000\n",
      "{'cluster': 81346, 'C-size': 31, 'label-counts': array([31])} {'metric': 0.4}\n",
      "edges: 90000\n",
      "{'cluster': 92025, 'C-size': 22, 'label-counts': array([22])} {'metric': 0.4}\n",
      "1\n",
      "Currently have this many samples: 100000\n",
      "!  seq_0 seq_19522 {'metric': 0.5920000000000001}  !\n",
      "Need to remove! Currently have this many samples: 100000\n",
      "After removal we have this many samples: 98410\n"
     ]
    }
   ],
   "source": [
    "train, test = train_test_split_homology(sdata.isel(_sequence=slice(0, int(1e5))), 'seq')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "EUGENe",
   "language": "python",
   "name": "eugene-dev"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
